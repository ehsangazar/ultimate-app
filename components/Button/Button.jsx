/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import { useTheme } from 'emotion-theming'

const Button = ({
  type = 'button',
  children,
  widthAll,
  disabled,
  buttonType = 'primary',
  loading,
}) => {
  const theme = useTheme()
  return (
    <div
      css={css`
        width: 100%;
      `}
    >
      <button
        css={css`
          width: ${widthAll ? '100%' : 'auto'};
          background-color: ${theme.colors[buttonType]};
          color: white;
          font-size: ${theme.typography.paragraph};
          text-decoration: none;
          padding: 16px 24px;
          border: ${theme.border[1]};
          border-radius: ${theme.borderRadius[1]};
          display: inline-block;
          position: relative;
          cursor: pointer;
          transition-duration: 0.4s;
          &:hover {
            box-shadow: ${theme.boxShadow.light};
            background-color: ${theme.colors.secondary};
          }
          &:disabled {
            background-color: ${theme.colors.backgroundColorSecondary};
            cursor: no-drop;
          }
        `}
        disabled={disabled || loading}
        type={type}
      >
        {loading && (
          <div
            css={css`
              height: 21px;
              width: 100%;
              position: relative;
              display: flex;
              align-items: center;
              justify-content: center;
            `}
          >
            <img
              alt="loading"
              css={css`
                width: 40px;
                position: absolute;
              `}
              src="assets/img/loading.svg"
            />
          </div>
        )}
        {!loading && children}
      </button>
    </div>
  )
}

export default Button
