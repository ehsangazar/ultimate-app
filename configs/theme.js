const theme = {
  colors: {
    primary: '#2980b9',
    secondary: '#f39c12',
    tertiary: '#e74c3c',
    linkPrimary: '#2c3e50',
    linkSecondary: '#f1c40f',
    white: '#ffffff',
    backgroundColorDefault: '#ffffff',
    backgroundColorPrimary: '#ecf0f1',
    backgroundColorSecondary: '#2c3e50',
  },
  typography: {
    paragraph: {
      fontSize: 1,
      lineHeight: 1.6,
      fontWeight: 300,
    },
    h1: {
      fontSize: 1.802,
      lineHeight: 1.6,
      fontWeight: 500,
    },
    h2: {
      fontSize: 1.602,
      lineHeight: 1.6,
      fontWeight: 500,
    },
    h3: {
      fontSize: 1.424,
      lineHeight: 1.6,
      fontWeight: 500,
    },
    h4: {
      fontSize: 1.266,
      lineHeight: 1.6,
      fontWeight: 500,
    },
    h5: {
      fontSize: 1.125,
      lineHeight: 1.6,
      fontWeight: 500,
    },
    small: {
      fontSize: 0.889,
      lineHeight: 1.6,
      fontWeight: 100,
    },
  },
  spaces: ['0px', '2px', '4px', '8px', '16px', '32px'],
  borderRadius: ['0px', '4px'],
  border: ['0px', '2px', '4px'],
  borderColor: {
    light: '#bdc3c7',
  },
  boxShadow: {
    light: '0 4px 16px rgba(0, 1, 31, 0.2)',
  },
}

export default theme
