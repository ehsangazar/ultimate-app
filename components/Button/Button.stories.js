import React from 'react'
import Button from './Button'

export default { title: 'Button' }

export const simple = () => <Button>کلیک کنید</Button>
export const simpleWidthAll = () => <Button widthAll>کلیک کنید</Button>
