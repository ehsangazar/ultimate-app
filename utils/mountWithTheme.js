import React from 'react'
import { ThemeProvider } from 'emotion-theming'
import renderer from 'react-test-renderer'
import theme from '../configs/theme'

const mountWithTheme = (children) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  )
}

export default mountWithTheme
