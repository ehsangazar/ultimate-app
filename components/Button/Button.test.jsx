import React from 'react'
import serializer from 'jest-emotion'
import { fireEvent, render } from '@testing-library/react'
import { ThemeProvider } from 'emotion-theming'
import mountWithTheme from '../../utils/mountWithTheme'
import Button from './Button'
import theme from '../../configs/theme'

expect.addSnapshotSerializer(serializer)

describe('Button', () => {
  test('Renders Correctly', () => {
    const component = mountWithTheme(<Button>Sample Button</Button>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Shows Loading state', () => {
    const component = mountWithTheme(
      <Button loading>Sample Button with loading</Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Shows disabled state', () => {
    const component = mountWithTheme(
      <Button disabled>Sample Button with disabled state</Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Shows buttonType secondary', () => {
    const component = mountWithTheme(
      <Button buttonType="secondary">Sample Button with secondary style</Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Renders in DOM correctly', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Button>Sample Button in DOM</Button>
      </ThemeProvider>
    )

    expect(getByText(/Sample Button/)).toBeTruthy()
  })
  test('OnClick Works Fine', () => {
    let variable = 1
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Button
          onClick={() => {
            variable += 1
          }}
        >
          Sample Button in DOM
        </Button>
      </ThemeProvider>
    )

    fireEvent.click(getByText(/Sample Button/))

    expect(variable).toBe(2)
  })
})
