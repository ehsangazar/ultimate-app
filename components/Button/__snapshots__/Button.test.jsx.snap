// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders Correctly 1`] = `
.emotion-1 {
  width: 100%;
}

.emotion-0 {
  width: auto;
  background-color: #2980b9;
  color: white;
  font-size: font-size:1px;
  line-height: 1.6;
  font-weight: 300;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 16px 24px;
  border: 2px;
  border-radius: 4px;
  display: inline-block;
  position: relative;
  cursor: pointer;
  -webkit-transition-duration: 0.4s;
  transition-duration: 0.4s;
}

.emotion-0:hover {
  box-shadow: 0 4px 16px rgba(0,1,31,0.2);
  background-color: #f39c12;
}

.emotion-0:disabled {
  background-color: #2c3e50;
  cursor: no-drop;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
    onClick={[Function]}
    type="button"
  >
    Sample Button
  </button>
</div>
`;

exports[`Button Shows Loading state 1`] = `
.emotion-3 {
  width: 100%;
}

.emotion-2 {
  width: auto;
  background-color: #2980b9;
  color: white;
  font-size: font-size:1px;
  line-height: 1.6;
  font-weight: 300;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 16px 24px;
  border: 2px;
  border-radius: 4px;
  display: inline-block;
  position: relative;
  cursor: pointer;
  -webkit-transition-duration: 0.4s;
  transition-duration: 0.4s;
}

.emotion-2:hover {
  box-shadow: 0 4px 16px rgba(0,1,31,0.2);
  background-color: #f39c12;
}

.emotion-2:disabled {
  background-color: #2c3e50;
  cursor: no-drop;
}

.emotion-1 {
  height: 21px;
  width: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  width: 40px;
  position: absolute;
}

<div
  className="emotion-3"
>
  <button
    className="emotion-2"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <div
      className="emotion-1"
    >
      <img
        alt="loading"
        className="emotion-0"
        src="assets/img/loading.svg"
      />
    </div>
  </button>
</div>
`;

exports[`Button Shows buttonType secondary 1`] = `
.emotion-1 {
  width: 100%;
}

.emotion-0 {
  width: auto;
  background-color: #f39c12;
  color: white;
  font-size: font-size:1px;
  line-height: 1.6;
  font-weight: 300;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 16px 24px;
  border: 2px;
  border-radius: 4px;
  display: inline-block;
  position: relative;
  cursor: pointer;
  -webkit-transition-duration: 0.4s;
  transition-duration: 0.4s;
}

.emotion-0:hover {
  box-shadow: 0 4px 16px rgba(0,1,31,0.2);
  background-color: #f39c12;
}

.emotion-0:disabled {
  background-color: #2c3e50;
  cursor: no-drop;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
    onClick={[Function]}
    type="button"
  >
    Sample Button with secondary style
  </button>
</div>
`;

exports[`Button Shows disabled state 1`] = `
.emotion-1 {
  width: 100%;
}

.emotion-0 {
  width: auto;
  background-color: #2980b9;
  color: white;
  font-size: font-size:1px;
  line-height: 1.6;
  font-weight: 300;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 16px 24px;
  border: 2px;
  border-radius: 4px;
  display: inline-block;
  position: relative;
  cursor: pointer;
  -webkit-transition-duration: 0.4s;
  transition-duration: 0.4s;
}

.emotion-0:hover {
  box-shadow: 0 4px 16px rgba(0,1,31,0.2);
  background-color: #f39c12;
}

.emotion-0:disabled {
  background-color: #2c3e50;
  cursor: no-drop;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Sample Button with disabled state
  </button>
</div>
`;
